Start autofocus and set FPS through CLI

roguh opened the pull request:
This is useful if you want autofocus on startup or on-demand.

Pass a command line flag to autofocus once or on an interval:

```
Usage of ./mtplvcap:
  -af-focus-now int
    	autofocus once on startup after X seconds, default=0 (disabled)
  -af-interval int
    	disable or change the autofocus interval in seconds, default=0 (disabled)
  -fps int
    	set frame limit, default=0 (disabled)
(existing flags omitted)
```

roguh committed:
New flags to start autofocus periodically with -af-interval, once with -af-focus-now, set FPS with -fps.

- New methods for LVServer: AFFocusNow AFFocusAfter
- New initialization parameters for LVServer
- Add new flags to README files
- Improve goroutine management and use a single function
- Improve user-facing argument names

puhitaku commented:
Hi, I'll review this PR after the command line flag is added. Ping me when it's done. Thanks!

roguh commented:
Hello, thank you. I'll let you know when this is configurable. I marked the PR as a draft too.


I was also thinking of a way to trigger auto focus through an event, perhaps by sending a signal to the process? I wonder if the web socket itself supports this? This could happen in a future PR though.

One of my cameras autofocuses very quickly, the other does not so I'd prefer to do it on-demand sometimes.

Thanks for the awesome project!

puhitaku commented:
> trigger auto focus through an event

Turning auto-focus on and off via WebSocket (that's what the browser calls on the controller page) should work for this case, although it is not documented. I currently have no plans to make a breaking change to this API in the future.

roguh commented:
Yes, that makes sense. For now, I'll just write my own script to send the autofocus event to the websocket somehow.

roguh commented:
Ok, this PR is ready for review. I ran `gofmt` and updated the README files too. I can exclude the new `.sh` script if needed.

roguh commented:
@puhitaku these autofocus improvements are ready for review.

I also have another PR in progress to add better reconnection capabilities if the USB device closes or if the websocket server closes: https://github.com/roguh/mtplvcap/pull/1 That's for issue #46

puhitaku commented:
Thanks for being passionate about improving mtplvcap. I'd like to classify each commit into this PR or another since the accumulated changes are getting too far from the original topic.

Please refrain from making any further additions from now on.

- dc949e2, 3627d4b, def73dd, 2f9f3e9, 306057f, 46e68cc, dce6949 : please squash them into one before proceeding to a code review.
- 110b50e, d03e29a, d273e16 : please remove them; the script is not part of mtplvcap
- bda74e4 : please leave it as-is. I like this change while this change is out of topic
- 457606f, dce6949 : please remove them; ~it's a shame that this kind of fundamental bug exists in the code, but I'd like to have a mock DSLR rather than keeping the server-only mode and having the dummy flag everywhere.~ EDIT: I saw #46 just now and I understood that this crash won't happen in the current master. Passing around the flag is a mistake I made in the past. I'll keep it as a TODO.

roguh commented:
I've squashed the relevant commits and removed the script, as well as the -server-only fix.

I'll keep the script in my own files, maybe I'll make a new repo for it sometime.

I don't need -server-only anymore because I'm using my smartphone as a sort of dummy MTP device that doesn't support any of the operations :)

> Thanks for being passionate about improving mtplvcap.

You're welcome! I've got some more ideas, but I'll make sure to follow the contributor's guide and open a couple issues so we can discuss.

puhitaku commented on mtp/server.go:
```diff
@@ -138,6 +143,24 @@ type InfoPayload struct {
 	Frame  []byte   `json:"frame"`
 }
 
+func (s *LVServer) AFFocusNow() {
+	log.LV.Info("AFFocusNow: focus now")
+	select {
+	case s.afNowChan <- true:
+	default:
+	}
+}
+
+func (s *LVServer) AFFocusAfter(after time.Duration) {
+	go func() {
+		select {
+		case <-time.After(after):
+			s.AFFocusNow()
+		case <-s.ctx.Done():
+		}
+	}()
+}
+
```
1. AFFocusNow is used only once. Let's merge it into the caller.
2. The entire contents of AFFocusAfter is wrapped by a `go func() {}()`. In this case, creating the new goroutine on the caller's side is cleaner and provides more choice in terms of concurrency (sync/async call etc.)

```suggestion
func (s *LVServer) AFFocusAfter(after time.Duration) {
	select {
	case <-time.After(after):
	case <-s.ctx.Done():
		return
	}

	select {
	case s.afNowChan <- true:
	case <-s.ctx.Done():
		return
	}
}

```

roguh replied:
done

puhitaku commented on main.go:
```diff
@@ -35,6 +35,9 @@ func main() {
 	vendorID := flag.String("vendor-id", "0x0", "VID of the camera to search (in hex), default=0x0 (all)")
 	productID := flag.String("product-id", "0x0", "PID of the camera to search (in hex), default=0x0 (all)")
 	maxResolution := flag.Bool("max-resolution", false, "change the resolution to the max (experimental)")
+	afInterval := flag.Int64("af-interval", 0, "disable or change the autofocus interval in seconds, default=0 (disabled)")
+	afFocusNow := flag.Int("af-focus-now", 0, "autofocus once on startup after X seconds, default=0 (disabled)")
```
Most camera users will understand what "AF" stands for, but I'm slightly pessimistic about it. Let's make them more human-friendly. `autofocus-on-startup` is very long, but follows the description's sentence.

```suggestion
	afInterval := flag.Int64("autofocus-interval", 0, "disable or change the autofocus interval in seconds, default=0 (disabled)")
	afFocusNow := flag.Int("autofocus-on-startup", 0, "autofocus once on startup after X seconds, default=0 (disabled)")
```

roguh replied:
done

puhitaku commented on README.md:
````diff
@@ -351,10 +351,16 @@ I strongly recommend you to compile by yourself for Linux distributions as Linux
 ```sh
 $ ./mtplvcap -help
 Usage of ./mtplvcap:
+  -af-focus-now int
+        autofocus once on startup after X seconds, default=0 (disabled)
+  -af-interval int
+        disable or change the autofocus interval in seconds, default=0 (disabled)
````
Related change to the argument.

```suggestion
  -autofocus-on-startup int
        autofocus once on startup after X seconds, default=0 (disabled)
  -autofocus-interval int
        disable or change the autofocus interval in seconds, default=0 (disabled)
```

roguh replied:
done

puhitaku commented on README.ja.md:
````diff
@@ -364,10 +364,16 @@ mtplvcapとOBSを組み合わせることで、NikonのカメラをHDMIキャプ
 ```sh
 $ ./mtplvcap -help
 Usage of ./mtplvcap:
+  -af-focus-now int
+        autofocus once on startup after X seconds, default=0 (disabled)
+  -af-interval int
+        disable or change the autofocus interval in seconds, default=0 (disabled)
````
Related change to the argument.

```suggestion
  -autofocus-on-startup int
        autofocus once on startup after X seconds, default=0 (disabled)
  -autofocus-interval int
        disable or change the autofocus interval in seconds, default=0 (disabled)
```

roguh replied:
done

puhitaku commented on main.go:
```diff
@@ -142,6 +145,10 @@ func main() {
 	log.Info("started")
 
+	if *afFocusNow > 0 {
+		lvs.AFFocusAfter(time.Duration(*afFocusNow) * time.Second)
```
Related change to the AFFocusAfter().

```suggestion
		go lvs.AFFocusAfter(time.Duration(*afFocusNow) * time.Second)
```

roguh replied:
done

roguh commented:
I'll implement these changes!

I admit I'm still learning Go, thank you for your patience.

roguh commented:
I have implemented the requested changes. Should I squash the commits again?

puhitaku commented:
@roguh Yes, please. Sorry for being very very late.

roguh commented:
No problem. I'll push a squashed commit soon :)

roguh commented:
Done! This is ready for review.

puhitaku reviewed (approved):
LGTM, thanks for the patient contribution!
